File: changelog.js
#!/usr/bin/env node

// get previous Tag
// tag new version
//change new previous
//tag
// commit chore(version):update changelog
//push


'use strict';

if (!Array.prototype.last){
  Array.prototype.last = function(){
    return this[this.length - 1];
  };
};

var fs = require('fs');
var util = require('util');
var child = require('child_process');

var q = require('qq');

var lite = false;
var GIT_LOG_CMD = 'git log --invert-grep --grep="%s" -E --format=%s %s..%s';
var GIT_TAG_CMD = 'git describe --tags --abbrev=0';
var GIT_TAG_DATE_CMD = 'git log -1 --format=%ai %s';
var HEADER_TPL = "\n<a name='%s'></a>\n# %s (%s)\n\n";
var HEADLESS_TPL = "";
var LINK_ISSUE_LITE = '[#%s]';
var LINK_FEATURE_LITE = "[%s]";
var LINK_COMMIT_LITE = '[%s]';
var LINK_ISSUE = '[#%s](' + process.env.ISSUE_TRACKER + '/%s)';
var LINK_FEATURE = "[%s](https://wiki.services.local/dosearchsite.action?spaceSearch=false&queryString='%s')";
  var LINK_COMMIT = '([%s](' + process.env.GIT_COMMIT_LINK + '/%s))';

  var EMPTY_COMPONENT = '$$';


  var warn = function() {
    console.error('WARNING:', util.format.apply(null, arguments));
  };


  var parseRawCommit = function(raw) {
    if (!raw) return null;

    var lines = raw.split('\n');
    var msg = {}, match;

    msg.hash = lines.shift();
    msg.subject = lines.shift();
    msg.closes = [];
    msg.breaks = [];

    lines.forEach(function(line) {
      match = line.match(/(?:Closes|Fixes|Features|closes|fixes|features)\s#?([A-Z0-9_\-]+)/);
      if (match) msg.closes.push(match[1]);
    });

    match = raw.match(/BREAKING CHANGE:([\s\S]*)/);

    if (match) {
      console.error("BREAK >%s<", match[1]);
      msg.breaking = match[1];
    }

    var featType= ["pods", "pod", "config", "conf", "feat", "refactor", "codereview"];
    msg.body = lines.join('\n');
    if(msg.subject.indexOf('Merge ')<0){
      match = msg.subject.match(/^(.*)\((.*)\)\s*\:\s*(.*)$/);
      if (!match || !match[1] || !match[3]) {
        match = msg.subject.match(/^\[([^ \]]*)\]\s*\[([^\]]*)\]\s*(.*)$/);
      }

      if (!match || !match[1] || !match[3]) {
        warn('Incorrect message: %s %s', msg.hash, msg.subject);
        msg.type='bof';
        msg.component='?';
        msg.subject= msg.subject+ " par *" +  authorCommit(msg.body) +"*";
        // console.log("\nHERE >>>>> %s \n>>>> %s\n ", msg.subject, msg.body);
      } else {
        msg.type = match[1].toLowerCase();
        var prefixe = "";
        if (featType.indexOf(msg.type)>-1) {
          prefixe = "[" + msg.type + "] - "
          msg.type="feat"
        }
        msg.component = match[2];
        msg.subject = prefixe + match[3];
      }
      return msg;
    } else
    return null;
  };
  var authorCommit = function(commit) {
    return commit.split('\n').last();
  };

  var linkToIssue = function(issue) {
    if (issue.match(/^[A-Z]+-[0-9]+$/)) {
      if (lite == true){
        return util.format(LINK_ISSUE_LITE, issue);
      } else
      return util.format(LINK_ISSUE, issue, issue);
    }
    if (lite == true){
      return util.format(LINK_FEATURE_LITE, issue);
    } else
    return util.format(LINK_FEATURE, issue, issue);
  };


  var linkToCommit = function(hash) {
    if (lite == true){
      return "";
    } else
    return util.format(LINK_COMMIT, hash.substr(0, 8), hash);
  };


  var currentDate = function(date) {
    var now;
    if (!date)
      now = new Date();
    else
      now = new Date(Date.parse(date));

    var pad = function(i) {
      return ('0' + i).substr(-2);
    };

    return util.format('%d-%s-%s', now.getFullYear(), pad(now.getMonth() + 1), pad(now.getDate()));
  };


  var printSection = function(stream, title, section, printCommitLinks) {
    
    printCommitLinks = printCommitLinks === undefined ? true : printCommitLinks;
    var components = Object.getOwnPropertyNames(section).sort();

    if (!components.length || components[0]==EMPTY_COMPONENT ){
      return;
    };

    stream.write(util.format('\n\n## %s', title));
    components.forEach(function(name) {
      var prefix = "\n  -";
      var nested = section[name].length > 1;

      if (name !== EMPTY_COMPONENT) {
        if (nested) {

          stream.write(util.format("\n  - **%s:**\n", name));
          prefix = "\n    -";
        } else {
          prefix = util.format("\n  - **%s:**", name);
        }
      }

      var doublon="";
      section[name].forEach(function(commit) {

        if (printCommitLinks) {
          if (doublon == commit.subject)
          {
            stream.write(linkToCommit(commit.hash));
          } else {
            stream.write(util.format("%s %s\n  %s", prefix, commit.subject, linkToCommit(commit.hash)));
          }
          if (commit.closes.length) {
            stream.write(",\n   " + commit.closes.map(linkToIssue).join(', '));
          }
        } else {
          if (doublon != commit.subject){
            stream.write(util.format("%s %s\n", prefix, commit.subject));
          }
        }
        doublon=commit.subject;
      });
    });
  };


  var readGitLog = function(grep, from, to) {
    var deferred = q.defer();

    // TODO(vojta): if it's slow, use spawn and stream it instead
    console.error(util.format(GIT_LOG_CMD, grep, '%H%n%B%n%an%n==END==', from, to));
    //console.log(child.exec("pwd; cat .git/config"));
    child.exec(util.format(GIT_LOG_CMD, grep, '%H%n%B%n%an%n==END==', from, to), function(code, stdout, stderr) {
      var commits = [];
      stdout.split('\n==END==\n').forEach(function(rawCommit) {
        var commit = parseRawCommit(rawCommit);
        if (commit) commits.push(commit);
      });

      deferred.resolve(commits);
    });

    return deferred.promise;
  };


  var writeChangelog = function(data, stream, commits, version, date) {
    var sections = {
      fix: {},
      feat: {},
      perf: {},
      bof: {},
      breaks: {}
    };

    sections.breaks[EMPTY_COMPONENT] = [];

    commits.forEach(function(commit) {
      var section = sections[commit.type];
      var component = commit.component || EMPTY_COMPONENT;

      if (section) {        
        section[component] = section[component] || [];
        section[component].push(commit);
      }

      if (commit.breaking) {
        sections.breaks[component] = sections.breaks[component] || [];
        sections.breaks[component].push({
          subject: util.format("due to %s,\n %s", linkToCommit(commit.hash), commit.breaking),
          hash: commit.hash,
          closes: []
        });
      }
    });

    if (version == "HEAD") {
      stream.write(util.format(HEADLESS_TPL, date));
    } else {
      stream.write(util.format(HEADER_TPL, version, version, currentDate(date)));
    }
  
    printSection(stream, 'Bug Fixes', sections.fix);
    printSection(stream, 'Features', sections.feat);
    printSection(stream, 'Performance Improvements', sections.perf);
    printSection(stream, 'Breaking Changes', sections.breaks, false);
    printSection(stream, 'non conforme', sections.bof);
    console.error("End ", data.length);
    stream.write("\n\n" + data);
  };


  var getPreviousTag = function() {
    var deferred = q.defer();
    child.exec(GIT_TAG_CMD, function(code, stdout, stderr) {
      if (code) deferred.reject('Cannot get the previous tag.');
      else deferred.resolve(stdout.replace('\n', ''));
    });
    return deferred.promise;
  };


  var generate = function(data, file, to, from) {
    var stream;


    stream = file ? fs.createWriteStream(file, { flags: 'w'}):process.stdout;

    var tagDate= child.execSync(util.format(GIT_TAG_DATE_CMD, to), function(error, stdout, stderr) {
      // command output is in stdout
      return stdout;
    });

    if(from==null){
      getPreviousTag().then(function(tag) {
        console.error('Reading git log since', tag);
        readGitLog('^Merge', tag, "HEAD").then(function(commits) {
          console.error('Parsed', commits.length, 'commits');
          console.error('Generating changelog to', file || 'stdout', '(', to, ')');
          //  console.error('>>>>>>',commits[0],'<<<<<')
          writeChangelog(data, stream, commits, to, tagDate);
        });
      });
    } else {
      console.error('Reading git log between %s and %s (%s)', from, to, tagDate.toString());
      readGitLog('^Merge', from, to).then(function(commits) {
        if (to=="HEAD"){
          commits[0].subject = commits[0].subject + " *Last*";
          tagDate = ""; // commits[0].subject + commits[0].subject+ commits[0].subject + "_ (" + authorCommit(commits[0].body) + ")"
        }
        console.error('Parsed', commits.length, 'commits');
        console.error('Generating changelog to', file || 'stdout', '(', to, ')');
        writeChangelog(data, stream, commits, to, tagDate);
      });
    };
    //  stream.write(data);
  }


  // publish for testing
  exports.parseRawCommit = parseRawCommit;
  exports.printSection = printSection;
  // hacky start if not run by jasmine :-D
  //console.error(child);
  if (process.argv.length > 5 && process.argv[5].indexOf("lite")>0){
  
    lite=true;
  }

  var file = process.argv[3];
console.error(process.argv[4])
  var chunk='';
  if (file){
    var streamOld = fs.createReadStream(file, {encoding: 'utf8'});

    streamOld.on('readable', function() {
      var buf;
      while ((buf = streamOld.read()) !== null) {
        chunk = chunk+buf;
      }
    });

    streamOld.on( 'end' ,  function ()  {
      console.error( 'Read completed successfully.' );
      generate(chunk, file, process.argv[2], process.argv[4]);
    });

  } else
  generate('', null, process.argv[2], process.argv[4]);

File: changelog.go
// Go port of changelog.js: builds a markdown changelog from the git log between two refs.
package main

import (
	"fmt"
	"os"
	"os/exec"
	"regexp"
	"sort"
	"strings"
	"time"
)

const emptyComponent = "$$"

var (
	reCloses      = regexp.MustCompile(`(?:closes|fixes|features)\s#?([a-z0-9_\-]+)`)
	reBreaking    = regexp.MustCompile(`BREAKING CHANGE:([\s\S]*)`)
	reConvention  = regexp.MustCompile(`^(.*)\((.*)\)\s*:\s*(.*)$`)
	reBracket     = regexp.MustCompile(`^\[([^ \]]*)\]\s*\[([^\]]*)\]\s*(.*)$`)
	reIssue       = regexp.MustCompile(`^[A-Z]+-[0-9]+$`)
	reVersionSufx = regexp.MustCompile(`-.*$`)
	featTypes     = []string{"feat", "pods", "pod", "config", "conf", "refactor", "codereview"}
)

type commit struct {
	hash, subject, body, typ, component, breaking string
	closes                                        []string
}

type generator struct {
	dir  string
	lite bool
	out  *strings.Builder
}

func gitOut(dir string, args ...string) (string, error) {
	cmd := exec.Command("git", args...)
	cmd.Dir = dir
	cmd.Stderr = os.Stderr
	out, err := cmd.Output()
	return string(out), err
}

// envJS returns an environment variable as the JS `process.env.NAME` string interpolation did:
// an unset variable becomes the literal "undefined".
func envJS(name string) string {
	if v, ok := os.LookupEnv(name); ok {
		return v
	}
	return "undefined"
}

func authorOf(body string) string {
	lines := strings.Split(body, "\n")
	return lines[len(lines)-1]
}

func parseRawCommit(raw string) *commit {
	if raw == "" {
		return nil
	}
	lines := strings.Split(raw, "\n")
	c := &commit{hash: lines[0]}
	lines = lines[1:]
	if len(lines) > 0 {
		c.subject = lines[0]
		lines = lines[1:]
	}
	for _, line := range lines {
		if m := reCloses.FindStringSubmatch(strings.ToLower(line)); m != nil {
			c.closes = append(c.closes, strings.ToUpper(m[1]))
		}
	}
	if m := reBreaking.FindStringSubmatch(raw); m != nil {
		fmt.Fprintf(os.Stderr, "BREAK >%s<\n", m[1])
		c.breaking = m[1]
	}
	c.body = strings.Join(lines, "\n")
	if strings.Contains(c.subject, "Merge ") {
		return nil
	}
	m := reConvention.FindStringSubmatch(c.subject)
	if m == nil || m[1] == "" || m[3] == "" {
		m = reBracket.FindStringSubmatch(c.subject)
	}
	if m == nil || m[1] == "" || m[3] == "" {
		fmt.Fprintf(os.Stderr, "WARNING: Incorrect message: %s %s\n", c.hash, c.subject)
		c.typ = "bof"
		c.component = "?"
		c.subject = c.subject + " par *" + authorOf(c.body) + "*"
		return c
	}
	c.typ = strings.ToLower(m[1])
	prefix := ""
	for i, t := range featTypes {
		if i > 0 && t == c.typ { // same quirk as the JS: indexOf(type) > 0
			prefix = "[" + c.typ + "] - "
			c.typ = "feat"
			break
		}
	}
	c.component = m[2]
	c.subject = prefix + m[3]
	return c
}

func (g *generator) linkToIssue(issue string) string {
	if reIssue.MatchString(issue) {
		if g.lite {
			return fmt.Sprintf("[#%s]", issue)
		}
		return fmt.Sprintf("[#%s](%s/%s)", issue, envJS("ISSUE_TRACKER"), issue)
	}
	if g.lite {
		return fmt.Sprintf("[%s]", issue)
	}
	return fmt.Sprintf("[%s](https://wiki.services.local/dosearchsite.action?spaceSearch=false&queryString='%s')", issue, issue)
}

func (g *generator) linkToCommit(hash string) string {
	if g.lite {
		return ""
	}
	short := hash
	if len(short) > 8 {
		short = short[:8]
	}
	return fmt.Sprintf("([%s](%s/%s))", short, envJS("GIT_COMMIT_LINK"), hash)
}

// currentDate formats a `git log --format=%ai` date as YYYY-MM-DD in local time (like the JS Date did).
func currentDate(date string) string {
	now := time.Now()
	if date = strings.TrimSpace(date); date != "" {
		if t, err := time.Parse("2006-01-02 15:04:05 -0700", date); err == nil {
			now = t.Local()
		}
	}
	return now.Format("2006-01-02")
}

func (g *generator) linksFor(closes []string) string {
	links := make([]string, len(closes))
	for i, c := range closes {
		links[i] = g.linkToIssue(c)
	}
	return ",\n   " + strings.Join(links, ", ")
}

func (g *generator) printSection(title string, section map[string][]*commit, printCommitLinks bool) {
	components := make([]string, 0, len(section))
	for name := range section {
		components = append(components, name)
	}
	sort.Strings(components)
	if len(components) == 0 || components[0] == emptyComponent {
		return
	}
	fmt.Fprintf(g.out, "\n\n## %s", title)
	for _, name := range components {
		prefix := "\n  -"
		nested := len(section[name]) > 1
		if name != emptyComponent {
			if nested {
				fmt.Fprintf(g.out, "\n  - **%s:**\n", name)
				prefix = "\n    -"
			} else {
				prefix = fmt.Sprintf("\n  - **%s:**", name)
			}
		}
		doublon := ""
		for _, c := range section[name] {
			if printCommitLinks {
				if doublon == c.subject {
					g.out.WriteString(g.linkToCommit(c.hash))
				} else {
					fmt.Fprintf(g.out, "%s %s\n  %s", prefix, c.subject, g.linkToCommit(c.hash))
				}
			} else if doublon != c.subject {
				fmt.Fprintf(g.out, "%s %s\n", prefix, c.subject)
			}
			if len(c.closes) > 0 {
				g.out.WriteString(g.linksFor(c.closes))
			}
			doublon = c.subject
		}
	}
}

func (g *generator) readGitLog(from, to string) ([]*commit, error) {
	raw, err := gitOut(g.dir, "log", "--invert-grep", "--grep=^Merge", "-E", "--format=%H%n%B%n%an%n==END==", from+".."+to)
	if err != nil {
		return nil, err
	}
	var commits []*commit
	for _, chunk := range strings.Split(raw, "\n==END==\n") {
		if c := parseRawCommit(chunk); c != nil {
			commits = append(commits, c)
		}
	}
	return commits, nil
}

func (g *generator) writeChangelog(data string, commits []*commit, version, date string) {
	sections := map[string]map[string][]*commit{
		"fix": {}, "feat": {}, "perf": {}, "bof": {}, "breaks": {emptyComponent: {}},
	}
	for _, c := range commits {
		component := c.component
		if component == "" {
			component = emptyComponent
		}
		if section, ok := sections[c.typ]; ok {
			section[component] = append(section[component], c)
		}
		if c.breaking != "" {
			sections["breaks"][component] = append(sections["breaks"][component], &commit{
				subject: fmt.Sprintf("due to %s,\n %s", g.linkToCommit(c.hash), c.breaking),
				hash:    c.hash,
			})
		}
	}
	if version == "HEAD" {
		g.out.WriteString(" ") // util.format("", "") in the JS produced a single space
	} else {
		fmt.Fprintf(g.out, "\n<a name='%s'></a>\n# %s (%s)\n\n", reVersionSufx.ReplaceAllString(version, ""), version, currentDate(date))
	}
	g.printSection("Bug Fixes", sections["fix"], true)
	g.printSection("Features", sections["feat"], true)
	g.printSection("Performance Improvements", sections["perf"], true)
	g.printSection("Breaking Changes", sections["breaks"], false)
	g.printSection("non conforme", sections["bof"], true)
	g.out.WriteString("\n\n" + data)
}

// generateChangelog mirrors `changelog.js <to> <file> <from> [--lite]` run in dir.
// With a file, the existing content is kept below the new section and the file is rewritten;
// otherwise the markdown is returned. An empty from falls back to the latest tag.
func generateChangelog(dir, to, file, from string, lite bool) (string, error) {
	g := &generator{dir: dir, lite: lite, out: &strings.Builder{}}
	data := ""
	if file != "" {
		b, err := os.ReadFile(file)
		if err != nil {
			return "", err
		}
		data = string(b)
	}
	tagDate, err := gitOut(dir, "log", "-1", "--format=%ai", to)
	if err != nil {
		return "", fmt.Errorf("cannot get date of %s: %w", to, err)
	}
	if from == "" {
		if from, err = gitOut(dir, "describe", "--tags", "--abbrev=0"); err != nil {
			return "", fmt.Errorf("cannot get the previous tag: %w", err)
		}
		from = strings.TrimSpace(from)
		to = "HEAD"
	}
	fmt.Fprintf(os.Stderr, "Reading git log between %s and %s (%s)\n", from, to, strings.TrimSpace(tagDate))
	commits, err := g.readGitLog(from, to)
	if err != nil {
		return "", err
	}
	if to == "HEAD" {
		// changelog.js crashed here (commits[0] undefined) and produced nothing: keep the file
		// untouched and return an empty changelog, but say so.
		if len(commits) == 0 {
			fmt.Fprintf(os.Stderr, "WARNING: no commits between %s and HEAD, empty changelog\n", from)
			return "", nil
		}
		commits[0].subject += " *Last*"
		tagDate = ""
	}
	fmt.Fprintf(os.Stderr, "Parsed %d commits\n", len(commits))
	g.writeChangelog(data, commits, to, tagDate)
	if file != "" {
		return "", os.WriteFile(file, []byte(g.out.String()), 0o644)
	}
	return g.out.String(), nil
}
